content/sip-55-coroutines.md

```diff
@@ -0,0 +1,965 @@
+The intersection type `C & D` indicates that a call to `pipe` 
+can be green only if both `tick` and `tock` are green but 
+must be red if either `tick` is or if `tock` is.
+In contrast to the `Task` type above, which requires one function 
+for each mix of pure or concurrent functions, 
+this color-transparent `pipe` coroutine covers all four cases.
+
+#### Code transformation for coroutine
+
+The essential feature of a coroutine is that, at some specific points,
+it can suspend its execution and save its local variables, so that 
+later in may restore those variables and resume its execution.
+Thus, we need to preserve the state of a suspended coroutine, but we cannot
+leave it in the stack of the running thread, because upon suspension the 
+thread may _shift_ to do something else and pop the coroutine's frame,
+The way to implement these features is to save the state of a suspended 
+coroutine execution, with its local variables, in a heap object.
+The essence of the code transformation is to split the code of the coroutine
+into a set of state transitions applied to a "frame" object.
+
+To do this, the code transformation has to: 1)
+identify the **suspension points** of the coroutine, 2)
+generate a `Frame` class to store the coroutine state 
+between suspensions, 3)
+generate a `runNextLeg` function to carry out the transition
+between the entry, exit, and suspension points, 4)
+generate an `init` method that builds the initial frame, 
+and 5) finally transform every call to the coroutine
+into a call to that `init` method. 
+Let us see these steps as applied to the `gcd` example.
+
+**1. Identify suspension points:** 
+Both `isZero` and `mod` have a suspension point at the call to `shift`.
+In the body of `gcd`, its suspension points are the call to `isZero`, to `mod`, 
+and the recursive call to `gcd`.
+Thus, the execution of a `gcd` may be suspended at one of these four states: 
+at the beginning of the coroutine, at the call to `isZero`, 
+at the call to `mod`, or at the call to  `gcd`. 
+
+**2) Generate a Frame class:**
+To indicate which of these states an execution is at, the `Frame` class 
+for the `gcd` has a `state` field. It also has two fields for the parameters
+`a` and `b` of `gcd`, and the fields `z`, `m` and `g` to store the 
+intermediate results of each call.
+The frame also has a `result` field, used to record errors or cancellations.
+Finally, the `completion` field is used to signal a parent context
+whenever the coroutine is suspended or ended. 
+The methods `invokeSuspend` and `create` can be used as a hook
+for the underlying launcher to handle the execution of the coroutine.
+
+```scala
+/**
+ * @param result Any current or intermediate result between invocations
+ * @param completion The launcher continuation, used for scheduling and returning results at launcher boundaries.
+ */
+class GcdFrame(
+  var state: Int, 
+  var a: Int, var b: Int, var z: Boolean, var m: Int, var g: Int, 
```

Comment:
What if we have many local varaibles ,will that be a problem?

Reply:
An object of the `Frame` class stores in heap the method's stack frame. Just as with the stack frame, the size is proportional to the number of local variables in the method, which is known at compile time. We could do some optimisations, similar to the ones the compiler does with stack frames. For instance, we could have two local variables with non-overlapping time-lines (span between declaration or assignment and use) share a same "position". 

Note that the focus of this SIP is 1) the syntax for coroutines, and 2) a possible implementation known to work in other languages. Such optimisations would be addressed down the line.